### PR TITLE
Fixed cleaning of recycled resources and some minor cleaning and bug fixes

### PR DESCRIPTION
**Pull request by emilwestergren**

### CODE REVIEW

File: source/draw/gpu/GpuImageBin.ooc [hunk 1]
```diff
@@ -14,63 +14,64 @@
...
-		this _uv = FreeArrayList<GpuImage> new()
-		this _yuv422 = FreeArrayList<GpuImage> new()
+		this _monochrome = VectorList<GpuImage> new()
+		this _bgr = VectorList<GpuImage> new()
+		this _bgra = VectorList<GpuImage> new()
+		this _uv = VectorList<GpuImage> new()
+		this _yuv422 = VectorList<GpuImage> new()
+	}
+	_cleanList: func (list: VectorList<GpuImage>) {
+		list apply(|image| image _recyclable = false)
```
Comment (davidhesselbom) [inline source/draw/gpu/GpuImageBin.ooc]: Did you check the `apply` call for memory leaks?

File: source/draw/gpu/GpuImageBin.ooc [hunk 2]
```diff
@@ -14,63 +14,64 @@
...
-		this _bgra = FreeArrayList<GpuImage> new()
-		this _uv = FreeArrayList<GpuImage> new()
-		this _yuv422 = FreeArrayList<GpuImage> new()
+		this _monochrome = VectorList<GpuImage> new()
+		this _bgr = VectorList<GpuImage> new()
+		this _bgra = VectorList<GpuImage> new()
+		this _uv = VectorList<GpuImage> new()
+		this _yuv422 = VectorList<GpuImage> new()
+	}
+	_cleanList: func (list: VectorList<GpuImage>) {
```
Comment (davidhesselbom) [inline source/draw/gpu/GpuImageBin.ooc]: This is a static function, yeah?

File: source/draw/gpu/GpuImageBin.ooc [hunk 3]
```diff
@@ -82,19 +83,19 @@ GpuImageBin: class {
...
 		this _mutex unlock()
 	}
-	_search: func (size: IntSize2D, arrayList: FreeArrayList<GpuImage>) -> GpuImage {
+	_search: func (size: IntSize2D, list: VectorList<GpuImage>) -> GpuImage {
 		result := null
-		index := 0
-		for (i in 0..arrayList size) {
-			image := arrayList[i]
+		index := -1
+		count := list count
```
Comment (davidhesselbom) [inline source/draw/gpu/GpuImageBin.ooc]: Why store `count` in a separate variable?

File: source/draw/gpu/GpuImageBin.ooc (hunk 1 above)
Comment (davidhesselbom) [inline source/draw/gpu/GpuImageBin.ooc]: Looks to me like this creates a Func, which you then have no way of freeing afterwards.

Comment (emilwestergren) [inline source/draw/gpu/GpuImageBin.ooc]: Shame, I like the syntax but I guess we can't use it unless we fix the generated C code.

Comment (davidhesselbom) [inline source/draw/gpu/GpuImageBin.ooc]: You could modify the `apply` function to have a Boolean argument that tells whether to free the Func when it's done applying stuff. Someone has to tell someone whether the Func should be freed. If you want the program itself to figure out when it's time to free something you've allocated, you use a garbage collector.

Comment (emilwestergren) [inline source/draw/gpu/GpuImageBin.ooc]: I'll just make a for loop instead, it's only two lines of code anyway.

Comment (davidhesselbom) [inline source/draw/gpu/GpuImageBin.ooc]: Thanks

Comment (emilwestergren) [inline source/draw/gpu/GpuImageBin.ooc]: Well, you're very welcome sir.

**Comment by emilwestergren** [comment]
Updated.

**Pull request merged**